ckanext/adfs/controller.py

```diff
@@ -0,0 +1,153 @@
+# encoding: utf-8
+
+import logging
+import ckan.lib.helpers as h
+import ckan.model as model
+import ckan.plugins.toolkit as toolkit
+import pylons
+import base64
+from validation import validate_saml
+from metadata import get_certificates, get_federation_metadata
+from extract import get_user_info
+import ckan.lib.base as base
+import ckan.logic as logic
+import ckan.lib.mailer as mailer
+from ckan.common import _, c, request
+from ckan.controllers.user import UserController
+
+
+log = logging.getLogger(__name__)
+render = base.render
+check_access = logic.check_access
+get_action = logic.get_action
+NotFound = logic.NotFound
+NotAuthorized = logic.NotAuthorized
+
+
+class ADFSRedirectController(toolkit.BaseController):
+    """
+    A custom home controller for receiving ADFS authorization responses.
+    """
+
+    def login(self):
+        """
+        Handle eggsmell request from the ADFS redirect_uri.
+        """
+        try:
+            eggsmell = pylons.request.POST.get('wresult')
+            if not eggsmell:
+                request_data = dict(pylons.request.POST)
+                eggsmell = base64.decodestring(request_data['SAMLResponse'])
+        except:
+            log.info('ADFS eggsmell')
+            log.info(dict(pylons.request.POST))
+        # We grab the metadata for each login because due to opaque
+        # bureaucracy and lack of communication the certificates can be
+        # changed. We looked into this and took made the call based upon lack
+        # of user problems and tech being under our control vs the (small
+        # amount of) latency from a network call per login attempt.
+        metadata = get_federation_metadata(pylons.config['adfs_metadata_url'])
+        x509_certificates = get_certificates(metadata)
+        if not validate_saml(eggsmell, x509_certificates):
+            raise ValueError('Invalid signature')
+        username, email, firstname, surname = get_user_info(eggsmell)
+
+        if not email:
+            log.error('Unable to login with ADFS')
+            log.error(eggsmell)
+            raise ValueError('No email returned with ADFS')
+
+        user = model.User.by_name(username)
```

Comment:
Is there a reason we're switching from toolkit to interact directly with User Model here and below?

Reply:
We're using the User model to create ADFS users. The toolkit won't work since the username validators won't let us make usernames based on email addresses.